### DISCUSSION TITLE
Is my layout achievable on Karabiner?

### DISCUSSION THREAD

**Discussion by argenkiwi** [evan-liu/karabiner.ts#151] [Q&A]
After reading some threads online it would seem karabiner.ts is one of the best abstractions over Karabiner Elements. Before I go into this rabbit hole, I wanted to get some thought from the community on whether I can implement [my layout](https://github.com/argenkiwi/kenkyo) with it or not. 

These are the main components of it:
1. Functions, numbers and symbols (fumbol) layer.
2. Extend layer (navigational keys like arrows, home, end, page up/down, backspace, delete, etc.)
3. Home row modifiers.

A while ago I attempted to use Goku to implement it and getting the layers to work was not that hard. Home row modifiers, however, were a different story. There are other details as well in terms of the keys used to access the layers:
- CapsLock is used for the extend layer and I would like it to keep its original functionality when tapped (not held). 
- The space bar is used for accessing the fumbol layer and it is important it only activates it when held and and another key is pressed and released (if it activates to soon it would introduce error while typing).

Thanks in advance for your feedback.

**Comment by evan-liu** [comment]
Hi @argenkiwi, thanks for your interest in using karabiner.ts. 

There is [an example of a home row modifier on Karabiner-Elements](https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to-if-held-down/#more-advanced-example).

Which in karabiner.ts would be something like: 

```ts
rule('Change the f key to the left shift key if held down').manipulators([
  map('f', 'optionalAny')
    .toIfAlone('f', {}, { halt: true })
    .toIfHeldDown('left_shift')
    .toDelayedAction([], toKey('f'))
    .parameters({
      "basic.to_delayed_action_delay_milliseconds": 500,
      "basic.to_if_held_down_threshold_milliseconds": 500
    })
]),
```

Can you try if it works for you? 

TBH, I didn't know about this `halt` + `toDelayedAction` trick. Will add some abstractions soon, e.g. Use them on [layers](https://karabiner.ts.evanliu.dev/rules/layer), and tap-held features.

**Reply by argenkiwi** [reply]
Thanks! I checked out that example. I observed that the shift function only becomes active after the timeout. Ideally it should activate as soon as another key is pressed and released so you don't need to be too precise with the timing. I guess it's a similar problem to that of using the spacebar to activate a layer.
